Add a default mapping from EventLoopGroup and ChannelFactory #6839

eolivelli opened the pull request:
Introduce a EventLoopGroup#defaultChannelFactory as an hint to build a ChannelFactory if not configured.
This way the Channel type will be automatically discovered for simple use-cases

Motivation:

see issue #6839 

Modification:

I have added a EventLoopGroup#defaultChannelFactory(boolean serverSide) method to let the EventLoopGroup describe the usual type of Channel it handles.

This way code will be simpler:

before the change:
```
 ServerBootstrap b = new ServerBootstrap();
            b.group(bossGroup, workerGroup)
                .channel(NetworkUtils.isEnableEpoolNative() ? EpollServerSocketChannel.class : NioServerSocketChannel.class)
                .childHandler(channelInitialized)
                .option(ChannelOption.SO_BACKLOG, 128);

```
After the change:
```
 ServerBootstrap b = new ServerBootstrap();
            b.group(bossGroup, workerGroup)
                .childHandler(channelInitialized)
                .option(ChannelOption.SO_BACKLOG, 128);

```

Result:

Fixes #6839 

I have some questions:
- I needed to change the 'validate' method in order to create the channelFactory if not given, this is not very good for me but I did not find another way
- this PR does not include new test cases, I will add them as soon as the general approach to the problem wil be approved (at the moment I have tested all netty based projects and all is running as expected)

normanmaurer commented on transport/src/main/java/io/netty/channel/EventLoopGroup.java:
```diff
@@ -49,4 +49,10 @@
+     * Return the defaults {@link ChannelFactory} for channels to be created in the context of this group
+     * @return a default channel factory or null if not provided
+     */
+    ChannelFactory defaultChannelFactory(boolean serverSide);
```
@eolivelli we can not do this for now as this would be a breaking change (adding a method to an interface

eolivelli replied:
@normanmaurer that you for your review

I understand this point.
Netty is compiled for Java6 (I think) and so defaults method are not supported.
Another way to implement this feature would be to create a global registry which maps EventLoopGroup class with the Channel class.
This registry will be filled in static {} blocks in EventLoopGroup subclasses

I can file a new PR in order to demostrate this alternative approach

netkins commented on transport-native-kqueue/src/main/java/io/netty/channel/kqueue/KQueueEventLoop.java:
```diff
@@ -15,8 +15,10 @@
+import io.netty.channel.ChannelFactory;
 import io.netty.channel.EventLoop;
 import io.netty.channel.EventLoopGroup;
+import io.netty.channel.ReflectiveChannelFactory;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ReflectiveChannelFactory'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on transport/src/main/java/io/netty/channel/embedded/EmbeddedEventLoop.java:
```diff
@@ -16,11 +16,13 @@
 import io.netty.channel.DefaultChannelPromise;
 import io.netty.channel.EventLoop;
 import io.netty.channel.EventLoopGroup;
+import io.netty.channel.ReflectiveChannelFactory;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ReflectiveChannelFactory'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on transport/src/main/java/io/netty/channel/DefaultEventLoop.java:
```diff
@@ -15,6 +15,8 @@
  */
 package io.netty.channel;
 
+import io.netty.channel.local.LocalChannel;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.local.LocalChannel'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.local.LocalChannel'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on transport/src/main/java/io/netty/channel/DefaultEventLoop.java:
```diff
@@ -15,6 +15,8 @@
 package io.netty.channel;
 
+import io.netty.channel.local.LocalChannel;
+import io.netty.channel.local.LocalServerChannel;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.local.LocalServerChannel'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.local.LocalServerChannel'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on transport/src/main/java/io/netty/channel/nio/NioEventLoop.java:
```diff
@@ -17,10 +17,14 @@
 
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelException;
+import io.netty.channel.ChannelFactoriesRegistry;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelFactoriesRegistry'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelFactoriesRegistry'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on transport/src/main/java/io/netty/channel/nio/NioEventLoop.java:
```diff
@@ -17,10 +17,14 @@
+import io.netty.channel.ChannelFactoriesRegistry;
 import io.netty.channel.EventLoop;
 import io.netty.channel.EventLoopException;
+import io.netty.channel.ReflectiveChannelFactory;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ReflectiveChannelFactory'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ReflectiveChannelFactory'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on transport/src/main/java/io/netty/channel/nio/NioEventLoop.java:
```diff
@@ -17,10 +17,14 @@
+import io.netty.channel.ReflectiveChannelFactory;
 import io.netty.channel.SelectStrategy;
 import io.netty.channel.SingleThreadEventLoop;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.socket.nio.NioServerSocketChannel'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.socket.nio.NioServerSocketChannel'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on transport/src/main/java/io/netty/channel/nio/NioEventLoop.java:
```diff
@@ -17,10 +17,14 @@
 import io.netty.channel.SelectStrategy;
 import io.netty.channel.SingleThreadEventLoop;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.socket.nio.NioSocketChannel'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.socket.nio.NioSocketChannel'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java:
```diff
@@ -0,0 +1,43 @@
+ * This class provides a default mapping between {@link EventLoopGroup} implementations and {@link ChannelFactory}
+ * implementations
+ */
+public abstract class ChannelFactoriesRegistry {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1118)

netkins replied:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1118)

netkins commented on transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java:
```diff
@@ -218,6 +220,14 @@ public B validate() {
+    private void ensureChannelFactory() {
+        if (channelFactory == null) {
+            channelFactory = ChannelFactoriesRegistry.getFactoryForEventLoopGroup(group.getClass(), isServerSide());
+            System.out.println("[DEBUG] channelFactory for " + group.getClass()
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

netkins replied:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

mp911de commented:
It would make sense to include `KQueue` as well.

eolivelli commented:
@mp911de I pushed the support for KQueue

@normanmaurer If the general approch is good I would like to add test cases.
Which would be the best place to put them ?
I think that as I am adding default socket channel configuration for Nio, Default, Epoll and KQueue at least I should add a test case in any of the projects

Scottmitch commented on transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java:
```diff
@@ -499,4 +509,6 @@ protected EventExecutor executor() {
         }
     }
+
+    protected abstract boolean isServerSide();
```
This method cannot be abstract because this class is public and it would break the API contract.

Scottmitch commented on transport/src/main/java/io/netty/channel/ThreadPerChannelEventLoop.java:
```diff
@@ -93,4 +93,5 @@ protected void deregister() {
         parent.activeChildren.remove(this);
         parent.idleChildren.add(this);
     }
+
```
Please revert back unrelated formatting changes

Scottmitch commented on transport/src/test/java/io/netty/channel/local/LocalChannelTest.java:
```diff
@@ -21,6 +21,7 @@
 import io.netty.buffer.Unpooled;
 import io.netty.channel.AbstractChannel;
 import io.netty.channel.Channel;
+import io.netty.channel.ChannelFactory;
```
Revert please

eolivelli commented:
@Scottmitch 
I have implemented only the cases for which I am sure about the most convenient default channel type.
My experience is only limited to "tcp" on Default/Nio/Epoll. I have added KQueue as requested.

Should I add some more channel types ?
Could you please give me an hint ?

eolivelli commented:
@Scottmitch  @normanmaurer 
does the current approch for you ?

@Scottmitch could you please give me some hint ? (see my answer to your change request)

rkapsi commented on transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java:
```diff
@@ -0,0 +1,45 @@
+
+    private ChannelFactoriesRegistry() { }
+
+    public static void registerFactoryForEventLoopGroup(Class<? extends EventLoopGroup> groupClass,
```
This is possibly not good for forward compatibility. Multiple people have filed issues over `EpollEventLoopGroup`, `NioEventLoopGroup` and so forth being concrete+final classes and basically asked them to be interfaces.

If we were to do that then this approach would no longer work. IMHO it'd be better to use `instanceof` in some shape or form instead of depending on hashCode/equals.

rkapsi commented on transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public abstract class ChannelFactoriesRegistry {
+
+    private static ConcurrentMap<Class<? extends EventLoopGroup>, ChannelFactory> clientSideFactories
```
`final`, upper case, just `Map` on the left hand side, I believe there is a factory method for concurrent maps in the `PlatformDependent` class or so.

eolivelli commented:
@rkapsi I tried to address your request. 
I have added a test case for the case of using EventLoopGroup subclasses

rkapsi commented on transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java:
```diff
@@ -218,6 +220,12 @@ public B validate() {
 
+    private void ensureChannelFactory() {
+        if (channelFactory == null) {
+            channelFactory = ChannelFactoriesRegistry.getFactoryForEventLoopGroup(group.getClass(), isServerSide());
```
Wonder if you can get away with `(this instanceof ServerBootstrap)` instead of the `isServerSide()` method.

eolivelli replied:
@rkapsi IMHO using "instanceof" in a parent class on 'this' is not a good object oriented design choice.
But I will change it if you confirm it is better for the rules of the project

fenik17 replied:
@eolivelli +1. A class reference to its subclasses is anti-pattern..

rkapsi replied:
How about this...

```java
class AbstractBootstrap {
  private void ensureChannelFactory() {
    if (chanelFactory == null) {
       chanelFactory = getDefaultChannelFactory();
    }
  }

  protected ChannelFactory getDefaultChannelFactory() {
    return null;
  }
}

class Bootstrap extends AbstractBootstrap {
  @Overrides
  protected ChannelFactory getDefaultChannelFactory() {
    return ChannelFactoriesRegistry.getFactoryForEventLoopGroup(group.getClass(), false);
  }
}

class ServerBootstrap extends AbstractBootstrap {
  @Overrides
  protected ChannelFactory getDefaultChannelFactory() {
    return ChannelFactoriesRegistry.getFactoryForEventLoopGroup(group.getClass(), true);
  }
}
```

eolivelli replied:
@rkapsi for me is OK I will update the patch soon, thanks

fenik17 replied:
@rkapsi this is better. One nit:
```protected abstract ChannelFactory getDefaultChannelFactory();```

eolivelli replied:
@fenik17 with cannot make it abstract because we would break existing subclasses

rkapsi replied:
I believe `abstract` would be an API breaking change.

fenik17 replied:
yep. I updated comment above. 
If `getDefaultChannelFactory()` returns `null`, this will lead to an runtime NPE in the users subclusses.

rkapsi replied:
Nah, should behave the same as right now saying something along the lines of ChannelFactory being missing.

You wouldn't want to return a client channel factory by default if the user's custom bootstrap is for server channels.

eolivelli replied:
@rkapsi @fenik17  getDefaultChannelFactory() by default will return null as it will return for type of EventLoopGroups which are not pre-registered.
I will update the patch soon

rkapsi commented on transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java:
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2012 The Netty Project
```
2017

rkapsi commented on transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java:
```diff
@@ -0,0 +1,75 @@
+        private final ChannelFactory serverSideFactory;
+
+        public ChannelFactoryDefaults(Class<? extends EventLoopGroup> eventLoopGroupType,
+            ChannelFactory clientSideFactory, ChannelFactory serverSideFactory) {
```
nit: I believe this needs an extra indent and I believe (unfortunately) is to actually align the args. Please double check and/or somebody can confirm.

rkapsi commented on transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java:
```diff
@@ -0,0 +1,75 @@
+    private ChannelFactoriesRegistry() { }
+
+    public static void registerFactoryForEventLoopGroup(Class<? extends EventLoopGroup> groupClass,
+        ChannelFactory factoryForClients, ChannelFactory factoryForServers) {
```
Same with these args

rkapsi commented on transport/src/test/java/io/netty/channel/ChannelFactoriesRegistryTest.java:
```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2015 The Netty Project
```
2017

rkapsi commented:
LGTM except for one aspect and something I've struggled myself with the API. The type of the `EventLoop[Group]` doesn't really indicate which "transport layer" the user intends to use. This impl. assumes the user wants to use TCP but it could be UDP, Unix Domain Socket or something else and I don't know if it's a good idea to make such default choice.

We did for a while the opposite and had an `EventLoopGroupManager` class with a `get(Class<? extends Channel> type)` method that provided appropriate `EventLoopGroup` instances for a given channel type and a custom wrapper around Bootstrap what would fill in the group if it was missing. It was OK but had its problems too when it came down to debugging things.

eolivelli commented:
@rkapsi thank you
I have fixed the indentation/copyright issues (I followed the example of PooledByteBufAllocator for the indentation).
I will commit the 'instanceof' change if needed, see my comments.

For the question about the support of connection less protocols or Unix Domain Socket....
I think that this feature makes the life simpler to new Netty users for the basic TCP usage, more advanced cases are still possible using the usual API.

eolivelli commented:
@rkapsi  @fenik17 pushed the refactor

eolivelli commented:
@rkapsi @fenik17 @Scottmitch @normanmaurer
any other comments ?

eolivelli commented:
Ping

fenik17 commented on transport/src/test/java/io/netty/channel/ChannelFactoriesRegistryTest.java:
```diff
@@ -0,0 +1,226 @@
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+            logger.info(String.format("Received message: %s", msg));
```
```logger.info("Received message: {}", msg);```

eolivelli replied:
@fenik17 done

eolivelli commented:
@rkapsi @fenik17 @Scottmitch @normanmaurer
any other comments ?

johnou commented on transport/src/main/java/io/netty/bootstrap/Bootstrap.java:
```diff
@@ -315,4 +316,9 @@ final SocketAddress remoteAddress() {
     }
+
+    @Override
+    protected io.netty.channel.ChannelFactory getDefaultChannelFactory() {
```
do we need the FQCN?

eolivelli replied:
@johnou yes, otherwise we are hitting a deprecated method

johnou commented on transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    private static final List<ChannelFactoryDefaults> DEFAULT_CHANNEL_FACTORIES
```
I think a COW would make more sense here so we can get rid of the sync.

eolivelli replied:
I think we can't. Otherwise we can't implement correctly registerFactoryForEventLoopGroup.

eolivelli commented:
@johnou I have answered to all of your comments.
Rebased to current 4.1 branch

normanmaurer commented:
I think we should revisit the way how we do this kind of stuff in netty 5, but for netty 4 this is too late. Let me close this.

eolivelli commented:
Sure
